internal/konnect/node_agent.go

```diff
@@ -188,10 +320,69 @@ func (a *NodeAgent) updateNodeLoop(ctx context.Context) {
 			a.Logger.Info("update node loop stopped", "message", err.Error())
 			return
 		case <-ticker.C:
-			err := a.updateNode()
+			err := a.updateNodes()
+			if err != nil {
+				a.Logger.Error(err, "failed to update nodes")
+			}
+		}
+	}
+}
+
+// GatewayClientGetter gets gateway instances from admin API clients.
+// It is used when gateway discovery is not enabled (`--kong-admin-svc` not set).
+// it will use `gateway_<address>` as node hostname, like `gateway_127.0.0.1`.
+type GatewayClientGetter struct {
+	logger          logr.Logger
+	clientsProvider dataplane.AdminAPIClientsProvider
+}
+
+var _ GatewayInstanceGetter = &GatewayClientGetter{}
+
+// NewGatewayClientGetter creates a GatewayClientGetter to get gateway instances from client provider.
+func NewGatewayClientGetter(logger logr.Logger, clientsProvider dataplane.AdminAPIClientsProvider) *GatewayClientGetter {
+	return &GatewayClientGetter{
+		logger:          logger.WithName("gateway-admin-api-getter"),
+		clientsProvider: clientsProvider,
+	}
+}
+
+// GetGatewayInstances gets gateway instances from currently available gateway API clients.
+func (p *GatewayClientGetter) GetGatewayInstances() ([]GatewayInstance, error) {
+	gatewayClients := p.clientsProvider.GatewayClients()
+	// TODO: get version of each kong gateway instance behind clients:
+	// https://github.com/Kong/kubernetes-ingress-controller/issues/3590
+	kongVersion := ""
+	if len(gatewayClients) != 0 {
+		v, err := gatewayClients[0].GetKongVersion(context.Background())
+		if err != nil {
+			p.logger.Error(err, "failed to get kong version")
+		} else {
+			kongVersion = v
+		}
+	}
+
+	gatewayInstances := make([]GatewayInstance, 0, len(gatewayClients))
+	for _, client := range gatewayClients {
+		var hostname string
+		podNN, ok := client.PodReference()
+		if ok {
+			hostname = podNN.String()
+		} else {
+			rootURL := client.BaseRootURL()
+			u, err := url.Parse(rootURL)
 			if err != nil {
-				a.Logger.Error(err, "failed to update node", "node_id", a.NodeID)
+				p.logger.Error(err, "failed to parse URL of gateway admin API from raw URL, skipping", "url", rootURL)
+				continue
 			}
+			// use "gateway_address" as hostname of konnect node.
+			// REVIEW: trim ports in addresses? like 127.0.0.1:8444 -> gateway_127.0.0.1
```

Comment:
I think that leaving ports might make sense as you can theoretically have two different instances listening on different ports of the same host. 👍 
```suggestion
```

Reply:
done